app/src/db/migrations/20201008142046_006-version-drafts.js

```diff
@@ -0,0 +1,110 @@
+      table.boolean('published').notNullable().defaultTo(false);
+    }))
+    .then(() => knex.schema.raw(`create or replace view form_vw as
+      SELECT DISTINCT ON (lower(f.name::text), fv.version, f.id) f.id,
+                                              f.name,
+                                              f.active,
+                                              f.description,
+                                              f.labels,
+                                              f."createdAt",
+                                              f."createdBy",
+                                              f."updatedAt",
+                                              f."updatedBy",
+                                              fv.id AS "formVersionId",
+                                              fv.version,
+                                              case when count(fip.code) = 0 then '{}'::varchar[] else array_agg(distinct(fip.code)) end AS "identityProviders",
+                                              case when count(ip.idp) = 0 then '{}'::varchar[] else array_agg(distinct(ip.idp)) end AS idps,
+                                              fv.published,
+                                              fv."updatedAt" as "versionUpdatedAt"
+    FROM form f
+            LEFT JOIN LATERAL (select v.* from form_version v where v."formId" = f.id order by case when v.published then 1 else 0 end desc, v.version desc limit 1) fv ON true
+            LEFT OUTER JOIN form_identity_provider fip ON f.id = fip."formId"
+            LEFT OUTER JOIN identity_provider ip ON fip.code::text = ip.code::text
+    GROUP BY f.id, f.name, f.active, f.description, f.labels, f."createdAt", f."createdBy", f."updatedAt",
+            f."updatedBy", fv.id, fv.version, fv.published, fv."updatedAt"
+    ORDER BY lower(f.name::text), fv.version DESC, f.id`))
+    .then(() => knex.schema.raw(`create or replace view user_form_access_vw as 
+select r."userId", u."keycloakId", u.username, u."fullName", u."firstName", u."lastName", u.email, r."formId", f.name as "formName", f.labels, f."identityProviders", f.idps, f.active, f."formVersionId", f.version, r.roles, p.permissions, f.published, f."versionUpdatedAt"
+from "user" u join user_form_roles_vw r on u.id = r."userId" join user_form_permissions_vw p on r."userId" = p."userId" and r."formId" = p."formId" join form_vw f on f.id = p."formId"
+order by lower(u."lastName"::text), lower(u."firstName"::text), lower(f.name::text)`));
+};
+
+exports.down = function(knex) {
+  // OK, this is really ugly.
+  // One cannot simply remove a column from a view.
+  // One must drop the view and then recreate; however, one must drop ALL the dependant objects too, then recreate them.
+  // Very lame.
+  // first, drop the view that uses the new column (and all objects that use that view)
+  // then we drop the column from the table
+  // then we re-create the view and all the views that depend on it.
```

Comment:
😬

Reply:
Haha, yup.  can only add columns to the end of the list (without dropping first).
This going to be a pain if we constantly update views... :(